File: java_benchmarks/src/main/java/com/cmpe220/benchmark/AbstractBenchmark.java
package com.cmpe220.benchmark;

abstract public class AbstractBenchmark {

    protected String category;
    protected String description;

    /**
    * Operations before the actual query
    * e.g. connection, initializations, etc.
    */
    protected void setupQuery() throws Exception {

    }

    /**
    * Operations after the query
    * e.g. cleanup, shutdown, etc.
    */
    protected void endQuery() throws Exception {

    }

    /**
    * Run the query to benchmark
    * Make sure that the query is synchronous!
    */
    abstract protected void runQuery() throws Exception;

    /**
    * @return milliseconds that the query took
    */
    public long benchmark() throws Exception {
        setupQuery();

        long startTime = System.currentTimeMillis();
        runQuery();
        long endTime = System.currentTimeMillis();

        endQuery();
        return endTime - startTime;
    }

    public String getCategory() {
        return category;
    }

    public String getDescription() {
        return description;
    }
}

File: nodejs_benchmarks/AbstractBenchmark.js

class AbstractBenchmark {

    constructor() {
        if (this.constructor == AbstractBenchmark) {
            throw new Error("AbstractBenchmark is abstract");
        }
        this.category = '';
        this.description = '';
    }

    /**
    * Operations before the actual query
    * e.g. connection, initializations, etc.
    */
    setupQuery() {

    }

    /**
    * Operations after the query
    * e.g. cleanup, shutdown, etc.
    */
    endQuery() {

    }

    /**
    * Operations before the actual query
    * e.g. connection, initializations, etc.
    */
    runQuery() {
        throw new Error("runQuery must be implemented");
    }

    /**
    * send milliseconds that the query took over TCP
    */
    benchmark() {
        this.setupQuery();

        var start_time = performance.now();
        this.runQuery();
        var end_time = performance.now();

        this.endQuery();
        // TODO: send return val over IPC with TCP
        return (end_time - start_time) / 1000000;
    }

    getCategory() {
        return this.category;
    }

    getDescription() {
        return this.description;
    }
}

module.exports = {
    AbstractBenchmark,
}
